api/pkg/user/service/service.go

```diff
@@ -0,0 +1,201 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package user
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/golang-jwt/jwt"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"github.com/tektoncd/hub/api/pkg/token"
+	"gorm.io/gorm"
+)
+
+type JWTScheme struct {
+	// Name is the scheme name defined in the design.
+	Name string
+	// Scopes holds a list of scopes for the scheme.
+	Scopes []string
+	// RequiredScopes holds a list of scopes which are required
+	// by the scheme. It is a subset of Scopes field.
+	RequiredScopes []string
+}
+
+// JWTAuth acts as middleware and implements the authorization logic for services.
+func (s *UserService) JWTAuth(handler http.HandlerFunc) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+
+		jwt := req.Header.Get("Authorization")
+
+		if len(jwt) > 6 && strings.ToUpper(jwt[0:7]) == "BEARER " {
+			jwt = jwt[7:]
+		}
+
+		claims, err := token.Verify(jwt, s.JwtConfig.SigningKey)
+		if err != nil {
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		scheme := &JWTScheme{
+			Name:   jwt,
+			Scopes: []string{"rating:read", "rating:write", "agent:create", "catalog:refresh", "config:refresh", "refresh:token"},
+		}
+
+		err = ValidateScopes(claims, scheme)
+		if err != nil {
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		userID, ok := claims["id"].(float64)
+		if !ok {
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// Set the userId in the headers
+		req.Header.Set("UserId", fmt.Sprintf("%v", userID))
+
+		handler.ServeHTTP(res, req)
+	}
+}
+
+// ValidateScopes takes user scopes and checks if it has the scope which
+// is required for accessing the api
+func ValidateScopes(claims jwt.MapClaims, scheme *JWTScheme) error {
+
+	if claims["scopes"] == nil {
+		return fmt.Errorf("invalid scopes")
+	}
+
+	scopes, ok := claims["scopes"].([]interface{})
+	if !ok {
+		return fmt.Errorf("invalid scopes")
+	}
+
+	scopesInToken := make([]string, len(scopes))
+	for _, scp := range scopes {
+		scopesInToken = append(scopesInToken, scp.(string))
+	}
+	if err := scheme.Validate(scopesInToken); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Validate returns a non-nil error if scopes does not contain all of
+// JWT scheme's required scopes.
+func (s *JWTScheme) Validate(scopes []string) error {
+	return validateScopes(s.RequiredScopes, scopes)
+}
+
+func validateScopes(expected, actual []string) error {
```

Comment:
do we have tests for func in this file?

Reply:
No there are no test for the func in this file